v1.8 backports 2020-11-13

twpayne opened the pull request:
* #13937 -- change default docker image repository from docker.io to quay.io (@aanm)
 * #13916 -- ipam/allocator: Fix nil check on node CIDR (@christarazi)
 * #13944 -- test: Increase timeout on privileged unit tests (@pchaigno)
 * #13969 -- ci: log in to docker in vagrant boxes (@nebril)
 * #13992 -- fqdn: Make maximum number of IPs per restored rule configurable (@jrajahalme)
 * #13912 -- ctmap: GC orphan SNAT entries (@brb)
 * #13991 -- dnsproxy: print total number of rules if too many (@kkourt)
 * #13994 -- Hubble-Relay: proxy metadata from originating client (@nathanjsweet)

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 13937 13916 13944 13969 13992 13912 13991 13994; do contrib/backporting/set-labels.py $pr done 1.8; done
```

brb committed:
ctmap: GC orphan SNAT entries

[ upstream commit c9810bf7b235ee371279c85f025535a86b1ea675 ]

This commit adds a mechanism to remove orphan SNAT entries.

We call an SNAT entry orphan if it does not have either a corresponding
CT entry or an SNAT entry in a reverse order. Both cases can happen due
to LRU eviction heuristics (both CT and NAT maps are of the LRU type).

The mechanism for the removal is based on the GC signaling in the
datapath. When the datapath SNAT routine fails to find a free mapping
after SNAT_SIGNAL_THRES attempts, it sends the signal via the perf ring
buffer. The consumer of the buffer is the daemon. After receiving the
signal it invokes the CT GC.

The newly implemented GC addition iterates over all SNAT entries and
checks whether a corresponding CT entry is found, and if not, it tries
to remove both SNAT entries (for original and reverse flows).

For now, I didn't add GC of orphan SNAT entries created by DSR to keep
complexity of changes as low as possible. This will come as a follow up.

Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Tom Payne <tom@isovalent.com>

christarazi committed:
ipam/allocator: Fix nil check on node CIDR

[ upstream commit 8c5c7375b8bac5a151f1c576e1b163fed4a2e8f3 ]

Previously, the nil check was useless because the variable is never
going to be nil, as it was initialized to an empty `nodeCIDRs` instance.
This made the "Unable to allocate node CIDR" error impossible to
surface.

This commit fixes this by making the variable not a pointer, but rather
a value, and reworking the further checks to ensure that the error in
allocating a node CIDR can be surfaced.

Fixes: ef6ecbdb61 ("add error log When ipam allocate nodecidr failure")
Fixes: https://github.com/cilium/cilium/pull/13299

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>

christarazi committed:
ipam/allocator: Add test for no allocator case

[ upstream commit 04660d75ce4314cdea8317ddc6db0b75bfc954ac ]

This commit adds a new test case for the podcidr package. While at it,
change the test function to verify the actual error, rather than
simplying checking if we hit an error.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>

christarazi committed:
ipam/allocator: Remove unnecessary fmt.Sprintf

[ upstream commit 05ef690c4534dd840afd085b73f13b0fcc1f55d5 ]

We only need to append a raw comma, so there's no need to format the
string. While we're at it, fix up the rest of the function for
readability.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>

christarazi committed:
ipam/allocator: Replace error with sentinel type

[ upstream commit 0071c54b57abfc883006122703f21dde66f05bd3 ]

This commit converts a plain error into a sentinel type to allow unit
testing to validate that this error is returned in the correct path.

Signed-off-by: Chris Tarazi <chris@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>

aanm committed:
change default docker image repository from docker.io to quay.io

[ upstream commit 434b056f7c2f25487068458684247063e2a52452 ]

With the recent limitations introduced by Docker, docker hub has been
rate limiting docker pulls on images for non-registered users. To avoid
this limitation we should switch the default container image
repositories to quay.io.

Signed-off-by: André Martins <andre@cilium.io>
Signed-off-by: Tom Payne <tom@isovalent.com>

pchaigno committed:
test: Increase timeout on privileged unit tests

[ upstream commit 35af73e2035154401c5907a8d3391bf45293c386 ]

Flakes on privileged unit tests with 'context deadline exceeded' are
becoming a bit more common. There seem to be a lot of variations [1] in
actual duration of those tests, with a long tail of high durations.

This commit doubles the timeout value in an effort to reduce flakes.

1 - https://github.com/cilium/cilium/issues/12862#issuecomment-674085848
Signed-off-by: Paul Chaignon <paul@cilium.io>
Signed-off-by: Tom Payne <tom@isovalent.com>

nebril committed:
ci: log in to docker in vagrant boxes

[ upstream commit 238262f15a681674c2c02b4650e311b947ab44fc ]

Signed-off-by: Maciej Kwiek <maciej@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>

brb committed:
ctmap: Iterate SNAT map once when doing GC

[ upstream commit 0c83c28963cdc2af5514b9707cae42d815afbdc1 ]

Previously, after receiving the signal from the datapath, we iterated
NAT map twice: first to compare against CT TCP map, second - against CT
any map.

Obviously, doing the iterations two times was inefficient. This commit
fixes that by passing both CT {TCP,any} maps to the NAT GC routine. This
allows the NAT GC to iterate once.

Suggested-by: Joe Stringer <joe@cilium.io>
Signed-off-by: Martynas Pumputis <m@lambda.lt>
Signed-off-by: Tom Payne <tom@isovalent.com>

kkourt committed:
dnsproxy: print total number of rules if too many

[ upstream commit b115c544aadc9999b26964687bfebd893798db5f ]

GetRules() will not process more than 1000 rules per port.
Print how many are the total rules in the message.

Signed-off-by: Kornilios Kourtis <kornilios@isovalent.com>
Signed-off-by: Tom Payne <tom@isovalent.com>

jrajahalme committed:
fqdn: Make maximum number of IPs per restored rule configurable

[ upstream commit 871e7e128b406f0753d48196c8049689d863d798 ]

Only count the number of IPs for each FQDN selector/rule when storing
rules for restoration, rather than ignoring later rules on a port
after previous rules have hit the maximum number of IPs.

Make the maximum number of IPs per restored rule configurable with the
new option '--tofqdns-max-ips-per-restored-rule' (default 1000).

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
Signed-off-by: Tom Payne <tom@isovalent.com>

nathanjsweet committed:
Hubble-Relay: proxy metadata from originating client

[ upstream commit df9f66e3ad6725d704edbd48b4283f9b81f3a3bf ]

PR-12865 attempted to accomplish this by proxying the
context object without modifying it, which is incorrect.
The incoming and outgoing metadata keys are actually
different and must be explicitly set in order for the
metadata to be properly proxied.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>
Signed-off-by: Tom Payne <tom@isovalent.com>

pchaigno reviewed (approved):
:+1: for my one line change :slightly_smiling_face:

christarazi reviewed (approved):
LGTM for my commits

brb reviewed (approved):
LGTM for my changes, thanks.

brb committed:
pkg/bpf: Wrap error to fix test failure

To fix the ctmap privileged test failure, the following needs to be
applied (otherwise, if ...; errors.Is(err, unix.ENOENT) is always false
in the PurgeOrphanNATEntries(); the change was introduced in v1.9 by
2283103):

Signed-off-by: Tom Payne <tom@isovalent.com>

aanm reviewed (approved):
My changes LGTM
